File: app/src/main/java/com/soobinpark/appcraft/readingbible/domain/model/VerseBookmark.kt
package com.soobinpark.appcraft.readingbible.domain.model

enum class VerseHighlight(
    val label: String,
) {
    None("없음"),
    Yellow("노랑"),
    Mint("민트"),
    Blue("파랑"),
    Pink("분홍"),
    Lavender("라벤더"),
    Orange("주황"),
}

data class VerseBookmark(
    val versionCode: String,
    val bookIndex: Int,
    val chapter: Int,
    val verse: Int,
    val text: String,
    val note: String = "",
    val isBookmarked: Boolean = true,
    val highlight: VerseHighlight = VerseHighlight.None,
    val isRead: Boolean = false,
    val createdAtMillis: Long,
) {
    val key: String = key(versionCode, bookIndex, chapter, verse)

    companion object {
        fun key(
            versionCode: String,
            bookIndex: Int,
            chapter: Int,
            verse: Int,
        ): String = listOf(versionCode, bookIndex, chapter, verse).joinToString(":")
    }
}

File: ios/Sources/Domain/Models/VerseBookmark.swift
import Foundation

enum VerseHighlight: String, CaseIterable, Codable, Sendable {
    case none
    case yellow
    case mint
    case blue
    case pink
    case lavender
    case orange

    var label: String {
        switch self {
        case .none: L10n.text("highlight.none")
        case .yellow: L10n.text("highlight.yellow")
        case .mint: L10n.text("highlight.mint")
        case .blue: L10n.text("highlight.blue")
        case .pink: L10n.text("highlight.pink")
        case .lavender: L10n.text("highlight.lavender")
        case .orange: L10n.text("highlight.orange")
        }
    }
}

struct VerseBookmark: Identifiable, Codable, Hashable, Sendable {
    var id: String { key }
    let versionCode: String
    let bookIndex: Int
    let chapter: Int
    let verse: Int
    let text: String
    var note: String
    var isBookmarked: Bool
    var highlight: VerseHighlight
    var isRead: Bool
    var createdAtMillis: Int64

    var key: String {
        BibleVerse.key(versionCode: versionCode, bookIndex: bookIndex, chapter: chapter, verse: verse)
    }

    var shouldPersist: Bool {
        isBookmarked || !note.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty || highlight != .none || isRead
    }

    func matches(filter: RecordFilter) -> Bool {
        filter.matches(self)
    }
}
